backends/memcheck/ceed-memcheck-serial.c

```diff
@@ -11,22 +11,42 @@
 
 #include "ceed-memcheck.h"
 
+//------------------------------------------------------------------------------
+// Backend Destroy
+//------------------------------------------------------------------------------
+static int CeedDestroy_Memcheck(Ceed ceed) {
+  Ceed_Memcheck *impl;
+
+  CeedCallBackend(CeedGetData(ceed, &impl));
+  CeedCallBackend(CeedFree(&impl));
+  return CEED_ERROR_SUCCESS;
+}
+
 //------------------------------------------------------------------------------
 // Backend Init
 //------------------------------------------------------------------------------
 static int CeedInit_Memcheck(const char *resource, Ceed ceed) {
-  Ceed ceed_ref;
+  Ceed           ceed_ref;
+  Ceed_Memcheck *impl;
 
   CeedCheck(!strcmp(resource, "/cpu/self/memcheck") || !strcmp(resource, "/cpu/self/memcheck/serial"), ceed, CEED_ERROR_BACKEND,
             "Valgrind Memcheck backend cannot use resource: %s", resource);
 
+  // Backend specific data
+  CeedCallBackend(CeedCalloc(1, &impl));
+  CeedCallBackend(CeedSetData(ceed, impl));
+
   // Create reference Ceed that implementation will be dispatched through unless overridden
   CeedCallBackend(CeedInit("/cpu/self/ref/serial", &ceed_ref));
   CeedCallBackend(CeedSetDelegate(ceed, ceed_ref));
 
   CeedCallBackend(CeedSetBackendFunction(ceed, "Ceed", ceed, "VectorCreate", CeedVectorCreate_Memcheck));
+  CeedCallBackend(CeedSetBackendFunction(ceed, "Ceed", ceed, "ElemRestrictionCreate", CeedElemRestrictionCreate_Memcheck));
+  CeedCallBackend(CeedSetBackendFunction(ceed, "Ceed", ceed, "ElemRestrictionCreateBlocked", CeedElemRestrictionCreate_Memcheck));
+  CeedCallBackend(CeedSetBackendFunction(ceed, "Ceed", ceed, "ElemRestrictionCreateAtPoints", CeedElemRestrictionCreate_Memcheck));
```

Comment:
Unfortunately, the way we split restrictions means we have only these 3 backend constructor functions as options. Maybe we should be splitting strided, standard, oriented, and curl oriented constructors. Keep the same signature for each though so the backends can recycle the same core function as they currently do?